Rule: ARIA required context role (ff89c9) (#255)
act-rules/act-rules.github.io: _rules/aria-required-context-role-ff89c9.md

---
id: ff89c9
name: ARIA required context role
rule_type: atomic
description: |
  This rule checks that an element with an explicit semantic role exists inside its required context.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - Accessibility tree
  - DOM Tree
acknowledgements:
  authors:
    - Wilco Fiers
    - Brian Bors
  previous_authros:
    - Anne Thyme Nørregaard
---

## Applicability

> The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has a [WAI-ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/) [explicit semantic role](#explicit-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role).

Reviewer: The reason for this is explained here: https://github.com/act-rules/act-rules.github.io/pull/255/files#diff-abef745b5106975263959c2fb4399b7cR32

**Note:** An example of an element that has a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) is `tab` that has `tablist` as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

> **Note:** An example of an element that has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role) is an `li` element that has `role="listitem"`. These elements are not applicable

Reviewer 2: ```suggestion
**Note:** An example of an element that has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role) is an `li` element that has `role="listitem"`. These elements are not applicable.
```

## Assumptions

If the [explicit semantic role](#explicit-role) on the target element is incorrectly used, and any relationships between elements are already programmatically determinable, failing this rule may not result in accessibility issues for users of assistive technologies, and it should then not be considered a failure under WCAG success criterion 1.3.1 Info and Relationships.

## Accessibility Support

- User agents do not all have the same accessibility tree. Particularly the method of deriving which element owns which other elements varies between browsers. This can lead to different results from this rule, depending on which accessibility tree is used as input.
- `aria-owns` has limited support in some user agents.

## Background

- [Understanding Success Criterion 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
- [Required Context Role](https://www.w3.org/TR/wai-aria-1.1/#scope)

## Test Cases

> ### Passed

Reviewer 2: Maybe add a Passed Example where both roles are implicit (`ul` + `li`).

Reviewer: This rule is not applicable to implicit roles.

#### Passed Example 1

> Element with [explicit semantic role](#explicit-role) `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) `list`, expressed as an [explicit semantic role](#explicit-role).

Reviewer: Unnecessary.

> Element with [explicit semantic role](#explicit-role) `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) `list`, expressed as an [explicit semantic role](#explicit-role).

Reviewer 2: Can we at least go for
```suggestion
The element with an [explicit semantic role](#explicit-role) of `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) of `list`, expressed as an [explicit semantic role](#explicit-role).
```
in order to make it a proper English sentence.

```html
<div role="list">
	<div role="listitem">List item 1</div>
</div>
```

#### Passed Example 2

Element with [explicit semantic role](#explicit-role) `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) `list`, through the [implicit semantic role](#implicit-role) of `ul`.

```html
<ul>
	<div role="listitem">List item 1</div>
</ul>
```

#### Passed Example 3

Element contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) even though it is not a direct child of the [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

```html
<div role="list">
	<div role="presentation">
		<div role="listitem">List item 1</div>
	</div>
</div>
```

#### Passed Example 4

> `aria-owns` used to give the target element the right [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

Reviewer 2: ```suggestion
The second `div` (`item`) is owned by the first one through the `aria-owns` attribute. Hence the target element has the right [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).
```

```html

> <div role="list" aria-owns="id1"></div>

Reviewer 2: ```suggestion
<div role="list" aria-owns="item"></div>
```

> <div id="id1" role="listitem">List item 1</div>

Reviewer 2: ```suggestion
<div id="item" role="listitem">List item 1</div>
```

```

#### Passed Example 5

> `aria-owns` trumps ownership by closest ancestor, giving the element with [explicit semantic role](#explicit-role) of `listitem` the correct [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

Reviewer 2: ```suggestion
The `aria-owns` attribute override normal DOM tree relationship. Thus, the innermost `div` element (`item`) is not owned by the intermediate one (with a `role="navigation"`) and has the correct [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).
```

```html

> <div role="list" aria-owns="id1">

Reviewer 2: ```suggestion
<div role="list" aria-owns="item">
```

	<div role="navigation">

> 		<div id="id1" role="listitem">List item 1</div>

Reviewer 2: ```suggestion
		<div id="item" role="listitem">List item 1</div>
```

	</div>
</div>
```

#### Passed Example 6

> Since implicit ownership can cross shadow boundaries, the element with the [explicit semantic role](#explicit-role) of `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) `list`.

Reviewer 2: ```suggestion
Since implicit ownership can cross shadow boundaries, the element with the [explicit semantic role](#explicit-role) of `listitem` is contained within its [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) of `list`.
```

```html
<div id="host" role="list"></div>

<script>
	const host = document.querySelector('#host')
	const root = host.attachShadow({ mode: 'open' })
	root.innerHTML = '<div role="listitem">List item 1</div>'
</script>
```

### Failed

#### Failed Example 1

> The `listitem` has no [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

Reviewer 2: ```suggestion
The `div` element with a`role="listitem"` has no [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).
```

```html
<div role="listitem">List item 1</div>
```

#### Failed Example 2

> The `listitem` is owned by the `tabpanel`, because it is the closest ancestor, but `tabpanel` is not the correct [context](https://www.w3.org/TR/wai-aria-1.1/#scope) for `listitem`.

Reviewer 2: ```suggestion
The innermost `div` element is [owned by](#owned-by) the intermediate one, but `tabpanel` is not the correct [context](https://www.w3.org/TR/wai-aria-1.1/#scope) for `listitem`.
```

```html
<div role="list">
	<div role="tabpanel">
		<div role="listitem">List item 1</div>
	</div>
</div>
```

#### Failed Example 3

> The `listitem` is owned by the `aria-label="menu"` div, rather than the `list`.

Reviewer 2: ```suggestion
The innermost `div` element is [owned by](#owned-by) the intermediate one, which does not have the correct [context](https://www.w3.org/TR/wai-aria-1.1/#scope) for `listitem`.
```

```html
<div role="list">

> 	<div aria-label="menu">

Reviewer 2: The `aria-label` is useless for this example, as far as I can tell.
Also, this example is pretty similar to Failed Example 2 and maybe not necessarily.

Reviewer: Adding aria-label forces this element into the accessibility tree. That causes the list to get disconnected from its listitems, so users can no longer get the size of the list, or navigate through listitems like they would if the div had role=presentation on it.

Reviewer 2: Good point. Thanks for the explanation.

		<div role="listitem">List item 1</div>
	</div>

> </div>

Reviewer 2: I'm not fully understanding why we want this to fail rather than pass (see also comment on `owned-by`).

Reviewer 2: Got better understanding after discussion. Thanks.

```

#### Failed Example 4

Since explicit ownership cannot cross shadow boundaries, the element with the [explicit semantic role](#explicit-role) of `listitem` does not have a [context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

```html
<div role="list" aria-owns="item"></div>

<div id="host"></div>

<script>
	const host = document.querySelector('#host')
	const root = host.attachShadow({ mode: 'open' })
	root.innerHTML = '<div id="item" role="listitem">List item 1</div>'
</script>
```

### Inapplicable

#### Inapplicable Example 1

There is no element with an [explicit semantic role](#explicit-role).

```html
<ul>
	<li>List item 1</li>
</ul>

> ```

Reviewer 2: I do not fully understand why we don't care about elements without explicit role.
Notably, I guess that
```html
<div role="tabpanel">
  <li>Foo</li>
</div>
```
is a pretty bad idea. This is currently inapplicable. Why don't we want to have that applicable (and failing)? What other problem would it cause elsewhere?

Reviewer: This will be caught by the ARIA required owned element rule #245. The second part of that is, in HTML, the rules have to be a little different. In HTML, this isn't OK:

```
<ul>  <div role="none">  <li>...</li>  </div>  </ul>
```

That requires logic that I think is better put into a separate rule. Definitely not saying that shouldn't be tested, just that adding that kind of logic to this rule would make this more complicated than having separate rules.

Reviewer 2: Hmm… As far as I understand #245 
```html
<div role="tablist">
  <div role="tab">Foo</div>
  <li>Bar</li>
</div>
```
will pass "ARIA required owned element" because the `tablist` has a `tab` child. And the `tab` will similarly pass this rule. And the `li` will be ignored by this rule (because it does not have an explicit role) and thus won't be caught by any rule currently.

Reviewer: We'll look at this in the aria required owned element rule. That's not part of this rule.

#### Inapplicable Example 2

> The `listitem` is not [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 2: ```suggestion
The `div` element is not [included in the accessibility tree](#included-in-the-accessibility-tree).
```

```html
<div role="listitem" style="display:none;">List item 1</div>
```

#### Inapplicable Example 3

> The `header` does not have a [required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) listed in WAI-ARIA 1.1.

Reviewer 2: ```suggestion
The `header` role does not have a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope).
```

```html
<div role="header" aria-level="1">Hello!</div>
<p>Welcome to my homepage!</p>
```

#### Inapplicable Example 4

> The `listitem` has an [explicit semantic role](#explicit-role), but it is identical to the [implicit semantic role](#implicit-role), making the element inapplicable.

Reviewer 2: ```suggestion
The `li` element has an [explicit semantic role](#explicit-role), but it is identical to its [implicit semantic role](#implicit-role), making the element inapplicable.
```

Reviewer 2: I'd still use
```suggestion
The `listitem` has an [explicit semantic role](#explicit-role), but it is identical to its [implicit semantic role](#implicit-role), making the element inapplicable.
```

```html
<ul>
	<li role="listitem">List item 1</li>
</ul>

> ```

Reviewer 2: I do not understand why we want that case to be inapplicable.
What is so special in having identical explicit and implicit role that make us discard this.

I kinda guess this is related to the fact that we also don't care about element without explicit roles, and this one essentially has no explicit role. Butt hen, I do not either fully understand why we don't care about elements without explicit role…

#### Inapplicable Example 5

> The `doc-biblioentry` has a role from the [Digital Publishing WAI-ARIA Module (DPUB ARIA) 1.0](https://www.w3.org/TR/dpub-aria-1.0/), not the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/), and it is therefore inapplicable.

Reviewer 2: ```suggestion
The `p` element has a role of `doc-biblioentry` which is from the [Digital Publishing WAI-ARIA Module (DPUB ARIA) 1.0](https://www.w3.org/TR/dpub-aria-1.0/), not the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/), and it is therefore inapplicable.
```

---
id: ff89c9
name: ARIA required context role
rule_type: atomic
description: | 
  This rule checks that a role exists inside its required context.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects: 
- DOM Tree
- CSS Styling
authors:
- Anne Thyme Nørregaard
---

## Applicability

> The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has an explicit [semantic role](#semantic-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an implicit semantic role that is identical to its explicit semantic role.

Reviewer: I think an example would make this more readable:

>  has an explicit [semantic role](#semantic-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) (i.e. `listitem` has `list` as its required context),

Author: Will try it out, but go for `tab` that has `tablist` as a required context, since most other roles (including `listitem`) has several possible context roles, which confuses the example.

Author: Example added as a note to match the other note that was already there to give an example on what it means that the implicit/explicit roles are different.

**Note:** An example of an element that has an implicit semantic role that is identical to its explicit semantic role is an `<li>` element that has `role="listitem"`. This rule is not applicable to elements like this.

## Expectation

> The target element is [owned by](#owned-by) an element that has a [semantic role](#semantic-role) that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element.

Reviewer: We should also allow roles that inherit from the required context. Romain came up with this one a few days ago. You could put this as an passed example:

```html
<div role="dictionary">
  <div role="listitem">Foo</div>
</div>
```

Author: @WilcoFiers, as I understand, what you are asking for is to change the applicability to something like: 

> The target element is [owned by](#owned-by) an element that has a [semantic role](#semantic-role) that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element, or that inherits from the one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element.

... where "inherits from a role" means "has the role in question listed as a [superclass role](https://www.w3.org/TR/wai-aria-1.1/#superclassrole) (and the roles that inherits can reversely been found listed as [subclass roles](https://www.w3.org/TR/wai-aria-1.1/#subclassroles) for the [superclass role](https://www.w3.org/TR/wai-aria-1.1/#superclassrole)."

I don't know if I agree with this. 
As long as we are working within the WAI-ARIA spec (and we are in this rule, according to the expectation that says "one of the WAI-ARIA required context roles), both superclass and subclass roles will be listed as possible required context roles for a role. See e.g. [`row` (role)](https://www.w3.org/TR/wai-aria-1.1/#row) that both lists [`grid` (role)](https://www.w3.org/TR/wai-aria-1.1/#grid) and [`table` (role)](https://www.w3.org/TR/wai-aria-1.1/#table), even though `grid` is a subclass of the `table` role.

I am not sure where you are getting the `role="dictionary"` from? It's not in the WAI-ARIA spec.
Maybe you are thinking about [`doc-glossary` role](https://www.w3.org/TR/dpub-aria-1.0/#doc-glossary) from DPUB-ARIA that has `term` and `definition` as required owned elements? Since that just has the superclass role of `landmark`, adding something on inheritance wouldn't help anything here. 
And then there is also the fact that this is in another standard than the one used in the expectation...

Author: @WilcoFiers, I have added a note on the superclass and subclass roles being listed as required context roles in WAI-ARIA. 
Is this, along with my previous comment, enough to handle your concern?

Reviewer: Hey Anne. Apologies, I messed up my example. What I was thinking of is `role="directory"`, or similarly, `role="feed"`. Neither of those are listed as required context roles for `listitem`.  I don't know if that's an oversight in the ARIA spec or not. DPUB-ARIA may also be something worth considering. It does something quite unusual, like this:

```html
<section role="doc-bibliography">
   <h1>Cited Works</h1>
   <div role="list">
      <p role="doc-biblioentry" id="b8cab5dd-bc24-459c-9858-7afa9da69b64">
         John Steinbeck, The Grapes of Wrath (New York: The Viking Press, 1939)
      </p>
   </div>
   …
</section>
```

In DPUB-ARIA, `doc-biblioentry` has a required context role of `doc-bibliography`. The way we've defined this rule, this example, straight out of DPUB-ARIA, would fail. My guess is that the DPUB-ARIA folks actually took the literal "owned elements" definition from ARIA (the one we established is wrong), and went with that. I'm not too sure what to do with that. Perhaps @rdeltour has any ideas?

Reviewer 3: It's always been my interpretation that the required context role included the stated role **and all its subclass roles**, but I'm no RDF expert. So yes, my take is that `listitem` can definitely be owned by `directory`. We should ask the ARIA folks for a clarification.

> My guess is that the DPUB-ARIA folks actually took the literal "owned elements" definition from ARIA (the one we established is wrong),

Can you please clarify (or give a pointer to) what do you mean by the "literal owned elements definition", and what was it that you established was wrong?

/cc @mattgarrish

Reviewer 3: FYI I just filed w3c/dpub-aria#15

Author: Thanks for getting back, @WilcoFiers and @rdeltour .

When looking in the WAI-ARIA spec for mentions of `directory`, I think it becomes pretty clear that it is missing in the list of required context roles for `listitem`.

E.g.: 

> `listitem` (role) 
> A single item in a list or directory.

It just seems like it has slipped from the table, because it is mentioned in paragraphs in different places.

Looking further into it, `list` also has the subclass of [`feed`](https://www.w3.org/TR/wai-aria-1.1/#feed), which is...

> A scrollable list of articles where scrolling may cause articles to be added to or removed from either end of the list.

In this case, I do not think `feed` should be considered an allowed context role for `listitem`, even though it is a subclass of `list`, which to me indicates that only the subclasses that are actually listed as required context roles should be considered.

Author: Suggestion from call between Wilco, Romain, Kasper and Anne:

"... that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element, or a DPUB subclass of one the required context roles."

Add note about "directory" being incorrect.

Author: Raised an issue for the WAI-ARIA group on the `directory` (role) missing as a required context role for `listitem`: https://github.com/w3c/aria/issues/1030

This rule assumes that the explicit [semantic role](#semantic-role) on the target element is used with the intention to comply to WCAG. If the explicit semantic role on the target element is incorrectly used, and any relationships between elements are already programmatically determinable, failing this rule might not result in accessibility issues for users of assistive technologies, and it should then not be considered a failure under WCAG success criterion 1.3.1 Info and Relationships.

## Accessibility Support

- This rule relies on assistive technologies to recognize which elements are [owned by](#owned-by) each other. This includes when the element is owned by another element that is an ancestor, but not a parent of the target element. Some assistive technologies do not accept these owned by relationships, unless workarounds are used.
Furthermore, `aria-owns` has limited support in some user agents.
- Some user agents and assistive technologies ignore empty elements, which means they are not presented to all users. However, since this is handled inconsitently across user agents and assistive technologies, this rule is applicable to empty elements. 

## Background

> - [Required Context Role](https://www.w3.org/TR/wai-aria-1.1/#scope)

Reviewer: Missing a link to understanding, also are there any techniques?

Author: Link to Understanding added. No techniques map in a sensible way.

## Test Cases

> ### Passed

Reviewer: I think you should put some content into these listitems so that they render something on the page for testers.

Author: Content added to all applicable elements, but a Passed Example 7 has been added without any content, to ensure that implementations also see empty elements as applicable.

```html
<ul>
    <div role="listitem"></div>
</ul>
```

#### Passed Example 3

Element contained within its required context role even though it is not a direct child of the context role.

```html
<div role="list">
    <div>
         <div>
             <div>
                <div role="listitem"></div>
             </div>
         </div>
    </div>
</div>
```

#### Passed Example 4

`aria-owns` used to give the target element the right context role.

```html
<div role="list" aria-owns="id1"></div>
<div id="id1" role="listitem"></div>
```

#### Passed Example 5

`aria-owns` trumps ownership by closest ancestor, giving the element with role of `listitem` the correct context role.

```html
<div role="list" aria-owns="id1">
  <div role="tabpanel">
    <div id="id1" role="listitem"></div>
   </div>
</div>
```

#### Passed Example 6

Since implicit ownership can cross shadow boundaries, the element with the semantic role of `listitem` is contained within its required context role `list`.

```html
<div id="host" role="list">
</div>

<script>
  const host = document.querySelector("#host");
  const root = host.attachShadow({ mode: "open" });

  root.innerHTML = `

>     <slot></slot>

Reviewer: This slot element here doesn't do anything. I'd remove it.

Author: `slot` element removed. I hope it's true that it didn't do anything...

**Note:** An exampple of an element that has a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) is `tab` that has `tablist` as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

**Note:** An example of an element that has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role) is an `<li>` element that has `role="listitem"`. This rule is not applicable to elements like this.

## Expectation

The target element is [owned by](#owned-by) an element that has a [semantic role](#semantic-role) that is one of the [WAI-ARIA required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) of the target element.

**Note:** The definition of [owned by](#owned-by) used in this rule is diverging from the definition of ["owned element" in WAI-ARIA](https://www.w3.org/TR/wai-aria-1.1/#dfn-owned-element). See more in the [owned by](#owned-by) definition.

## Assumptions

> This rule assumes that the explicit [semantic role](#semantic-role) on the target element is used with the intention to comply to WCAG. If the explicit semantic role on the target element is incorrectly used, and any relationships between elements are already programmatically determinable, failing this rule might not result in accessibility issues for users of assistive technologies, and it should then not be considered a failure under WCAG success criterion 1.3.1 Info and Relationships.

Reviewer 4: need to move this also to the explicit semantic role definition

  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects: 
- DOM Tree
- CSS Styling
authors:
- Anne Thyme Nørregaard
---

## Applicability

The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has a [WAI-ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/) [explicit semantic role](#explicit-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role).

**Note:** An example of an element that has a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) is `tab` that has `tablist` as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

**Note:** An example of an element that has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role) is an `<li>` element that has `role="listitem"`. This rule is not applicable to elements like this.

> **Note:** The applicability of this rule is limited to only the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/), since inheritance of [required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) to [subclass roles](https://www.w3.org/TR/wai-aria-1.1/#subclassroles) seems to work differently within the [Digital Publishing WAI-ARIA Module 1.0](https://www.w3.org/TR/dpub-aria-1.0/) extension than within the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/) itself. The [WAI-ARIA Graphics Module](https://www.w3.org/TR/graphics-aria-1.0/) does not include any [required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope).

Reviewer 3: the thing is, the intent wasn't for DPUB to deviate from the ARIA model. Could we say perhaps something along the lines of:
> there are unresolved issues with how DPUB ARIA 1.0 uses role inheritance to define the required context, which makes it deviate from the model defined in ARIA 1.1.

> **Note:** The applicability of this rule is limited to only the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/), since inheritance of [required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope) to [subclass roles](https://www.w3.org/TR/wai-aria-1.1/#subclassroles) seems to work differently within the [Digital Publishing WAI-ARIA Module 1.0](https://www.w3.org/TR/dpub-aria-1.0/) extension than within the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/) itself. The [WAI-ARIA Graphics Module](https://www.w3.org/TR/graphics-aria-1.0/) does not include any [required context roles](https://www.w3.org/TR/wai-aria-1.1/#scope).

Author: Fixed

**Note:** For [`listitem` (role)](https://www.w3.org/TR/wai-aria-1.1/#listitem) the [`directory` (role)](https://www.w3.org/TR/wai-aria-1.1/#directory) seems to be missing as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) in the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1). [An issue for this has been filed](https://github.com/w3c/aria/issues/1030) with the Accessible Rich Internet Applications Working Group. 

> 

Author: Fixed

---
id: ff89c9
name: ARIA required context role
rule_type: atomic
description: | 

>   This rule checks that a role exists inside its required context.

Author: Fixed

accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects: 
- DOM Tree
- CSS Styling
authors:
- Anne Thyme Nørregaard
---

## Applicability

> The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has a [WAI-ARIA 1.1](https://www.w3.org/TR/wai-aria-1.1/) [explicit semantic role](#explicit-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role).

Author: Fixed

**Note:** For [`listitem` (role)](https://www.w3.org/TR/wai-aria-1.1/#listitem) the [`directory` (role)](https://www.w3.org/TR/wai-aria-1.1/#directory) seems to be missing as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) in the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1). [An issue for this has been filed](https://github.com/w3c/aria/issues/1030) with the Accessible Rich Internet Applications Working Group. 

## Assumptions

> This rule assumes that the [explicit semantic role](#explicit-role) on the target element is used with the intention to comply to WCAG. If the [explicit semantic role](#explicit-role) on the target element is incorrectly used, and any relationships between elements are already programmatically determinable, failing this rule might not result in accessibility issues for users of assistive technologies, and it should then not be considered a failure under WCAG success criterion 1.3.1 Info and Relationships.

Author: Fixed

## Accessibility Support

> - This rule relies on assistive technologies to recognize which elements are [owned by](#owned-by) each other. This includes when the element is owned by another element that is an ancestor, but not a parent of the target element. Some assistive technologies do not accept these owned by relationships, unless workarounds are used.

Reviewer: I'd rephrase this. You're essentially saying that elements own each other. That's not right. I would say replace "recognise which elements are owned by each other" with "consistently identify which element [owns] which other element."

> Furthermore, `aria-owns` has limited support in some user agents.

Reviewer: Remove the line-break.

Author: Fixed

## Background

- [Understanding Success Criterion 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
- [Required Context Role](https://www.w3.org/TR/wai-aria-1.1/#scope)

## Test Cases

### Passed

#### Passed Example 1

Element with role `listitem` is contained within its required context role `list`, expressed as an explicit role.

```html
<div role="list">
    <div role="listitem">List item 1</div>
</div>
```

#### Passed Example 2

Element with role `listitem` is contained within its required context role `list`, through the implicit role of `ul`.

```html
<ul>
    <div role="listitem">List item 1</div>
</ul>
```

#### Passed Example 3

Element contained within its required context role even though it is not a direct child of the context role.

```html
<div role="list">
    <div>
         <div>
             <div>
                <div role="listitem">List item 1</div>
             </div>
         </div>
    </div>
</div>
```

#### Passed Example 4

`aria-owns` used to give the target element the right context role.

```html
<div role="list" aria-owns="id1"></div>
<div id="id1" role="listitem">List item 1</div>
```

> #### Passed Example 5

Reviewer: This example wouldn't pass SC 1.3.1, can you fix that?

Author: @WilcoFiers, can you be more explicit as to what exactly it is you think wouldn't pass SC 1.3.1 in this test case?

Author: Agreed to put in landmark instead

<div role="list" aria-owns="id1">
  <div role="tabpanel">
    <div id="id1" role="listitem">List item 1</div>
   </div>
</div>
```

#### Passed Example 6

Since implicit ownership can cross shadow boundaries, the element with the semantic role of `listitem` is contained within its required context role `list`.

```html
<div id="host" role="list">
</div>

<script>
  const host = document.querySelector("#host");
  const root = host.attachShadow({ mode: "open" });

  root.innerHTML = `
    <div role="listitem">List item 1</div>
  `;
</script>
```
#### Passed Example 7

Element with role `listitem` is contained within its required context role `list`, expressed as an explicit role. Even though the element with role `listitem` is empty, it is still applicable to this rule.

```html
<div role="list">
    <div role="listitem"></div>
</div>
```

### Failed

#### Failed Example 1

No context role.

```html
<div role="listitem">List item 1</div>
```

#### Failed Example 2

Wrong context role.

```html
<div role="tablist">
    <div role="listitem">List item 1</div>
</div>
```

#### Failed Example 3

Element not contained within its required context role.

```html
<div role="list"></div>

>     <div role="listitem">List item 1</div>

Reviewer: This shouldn't indent. Can you please run prettier on this rule before you merge it?

Author: Fixed in this and one other test case

```

#### Failed Example 4

Element with role `listitem` has a closer ancestor, that is included in the accessibility tree, than the role `list` that should have been its context role.

```html
<div role="list">
    <div aria-label="menu">
         <div role="listitem">List item 1</div>
    </div>
</div>
```

**Note:** This test case follows the definition of [owned by](#owned-by) used in this rule. If implemented differently, this definition could cause differences in outcome for this test case. 

#### Failed Example 5

> Element with role `listitem` has a closer ancestor, that is included in the accessibility tree, than the role `list` that should have been its context role.

Reviewer: Should we just say that `listitem` is owned by the `tabpanel`, because it is a closer ancestor?

Author: Fixed

```html
<div role="list">
  <div role="tabpanel">
    <div role="listitem">List item 1</div>
   </div>

> </div>

Reviewer: Spacing is different from other TCs, please run prettier.

```

#### Failed Example 6

> The element with the semantic role of `listitem` is [owned by](#owned-by) the first element that references it through `aria-owns`, which results in the wrong context role.

Reviewer: Should "semantic role" be linked?

Author: Fixed

```html
<div role="tabpanel" aria-owns="id1">
  <div role="list" aria-owns="id1">
    <div id="id1" role="listitem">List item 1</div>
   </div>
</div>
```

**Note:** This test case follows the definition of [owned by](#owned-by) used in this rule. If implemented differently, this definition could cause differences in outcome for this test case.

#### Failed Example 7

Since explicit ownership cannot cross shadow boundaries, the element with the semantic role of `listitem` does not have a context role.

```html
<div role="list" aria-owns="item">
</div>

<div id="host"></div>

<script>
  const host = document.querySelector("#host");
  const root = host.attachShadow({ mode: "open" });

  root.innerHTML = `
    <div id="item" role="listitem">List item 1</div>
  `;
</script>
```

### Inapplicable

#### Inapplicable Example 1

Element does not have an explicit semantic role.

```html

> <ul>List item 1</ul>

Reviewer: Can you have this not fail SC 1.3.1?

```

#### Inapplicable example 2

Element is not exposed to assistive technologies.

```html
<div role="listitem" aria-hidden="true">List item 1</div>
```

#### Inapplicable Example 3

Role does not have any required context roles listed in WAI-ARIA spec.

```html
<div role="radio">Radio button 1</div>
```

#### Inapplicable Example 4

Element is not exposed to assistive technologies, since ancestor has `aria-hidden` attribute with value set to `true`.

```html
<div role="list" aria-hidden="true">
    <div role="listitem">List item 1</div>
</div>
```
#### Inapplicable Example 5

Element has an explicit semantic role, but it is identical to the implicit semantic role, making the element inapplicable.

```html
<li role="listitem">List item 1</li>
```

#### Inapplicable Example 6

Element has a role from the [Digital Publishing WAI-ARIA Module 1.0](https://www.w3.org/TR/dpub-aria-1.0/), not the [WAI-ARIA 1.1 Recommendation](https://www.w3.org/TR/wai-aria-1.1/), and it is therefore inapplicable.

```html

> <div role="doc-bibliography">Item 1<div>

Reviewer: Maybe include example 6 from DPUB-ARIA here? `doc-bibliography` doesn't have a required context role, so that's not a particularly meaningful example from DPUB. `doc-biblioentry` does though, so that might be worth including.

Author: Fixed

## Applicability

> The rule applies to any HTML or SVG element that is [included in the accessibility tree](#included-in-the-accessibility-tree) and has an explicit [semantic role](#semantic-role) with a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope), except if the element has an [implicit semantic role](#implicit-role) that is identical to its [explicit semantic role](#explicit-role).

Reviewer 4: You missed one instance of explicit semantic role

> **Note:** An exampple of an element that has a [WAI-ARIA required context role](https://www.w3.org/TR/wai-aria-1.1/#scope) is `tab` that has `tablist` as a [required context role](https://www.w3.org/TR/wai-aria-1.1/#scope).

Reviewer 4: exampple > example
